Merge .NET Standard projects into main solution

rprouse opened the pull request:
See #388 

**Please note, this update requires building with Visual Studio 2017** Previously, you could build on earlier versions of Visual Studio except for the .NET Standard version of the framework which required 2017.

Checkpointing my work and posting a PR so people are aware this is coming. At this point, it converts the main solution to use the new `csproj` format without any other changes. Still outstanding;

- [ ] Delete the .NET Standard solution
- [ ] Merge and multi-target the .NET Standard projects into their .NET Framework counterparts
- [ ] Update the NuGet packages to be multi-targeted. This will require the engine to look one directory higher when searching for extensions as a NuGet package. (Once we are capable of multi-targeting extensions, this logic will need to be revisited once again)
- [x] Update Cake to build and package the new projects

On a side-note, working with the new `csproj` format makes me happy. Best change Microsoft has made in awhile 😄

rprouse commented on src/NUnitConsole/nunit3-console/Options.cs:
```diff
@@ -158,13 +158,7 @@
 using NDesk.Options;
 #endif
 
-#if NUNIT_CONSOLE || NUNITLITE || NUNIT_ENGINE
```
This is the only real change in this file. I edited in VSCode, so it dropped the trailing spaces on save. Sorry.

jnm2 replied:
Sometimes I wonder how much worse our lives would really be if somebody submitted a whitespace consistency PR so we never had to deal with this again? 🤔

rprouse commented on src/NUnitEngine/mock-assembly/mock-assembly.csproj:
```diff
@@ -1,114 +1,30 @@
-    <PostBuildEvent>
-    </PostBuildEvent>
-  </PropertyGroup>
+  <ItemGroup>
```
@jnm2 do you remember what the fix is to prevent these from getting re-added in 15.5.6?

jnm2 replied:
The fix is 1) to use NUnit 3.10 which has [this .props](https://github.com/nunit/nunit/blob/a43bfb3481528f55f8a714e757a6f3d0395b6bff/nuget/framework/NUnit.props#L4) and 2) The VSTest team accidentally broke this in 15.6 so the GUID always shows up, but [if you use 15.7](https://github.com/Microsoft/vstest/issues/472#issuecomment-371211414) (currently preview) it'll go back to working.

rprouse replied:
Thanks, I was planning on updating to 3.10.1 in the next pass. By coincidence, I got a notification for the VSTest issue this morning 😄

rprouse commented on src/NUnitEngine/nunit.engine.tests/Services/ResultWriters/NUnit2XmlResultWriterTests.cs:
```diff
@@ -1,340 +0,0 @@
-﻿// ***********************************************************************
```
These two files weren't referenced in the tests and should have been deleted when the result writer was moved out to an extension.

ChrisMaddock commented:
Can't wait to have this in place! 

Rob - how do you feel about first merging a PR with just the existing solution converted to the new csproj - and then handling integrating the existing .NET Standard engine in a second PR?

My thinking: having a multi-targetable solution would be a help for the multi-targeted extensions work (for nunit.engine.api), and anything dotnet nunit-y we may want to look at could also be unblocked. Merging the current two engines, we already know is going to have a few problems, and be a bit of a messier job.

That said - I'm unlikely to get back to the extensions work for the next couple of weeks, so it may be that we won't have an issue.

rprouse commented:
I am okay with taking it in stages. I'll take a look at getting the packaging and cake build working again then see what it looks like.

rprouse commented:
Updated so that cake is building and packaging on my machine. AppVeyor failed because it was using the VS2015 image. I updated to the VS2017 image and am rebuilding. Will need to make the same changes that we made to the nunit repo to get building on Travis.

rprouse commented:
AppVeyor is now failing with the following test failure. Not sure why it is failing there, but not on my machine,

```
1) Failed : NUnit.Engine.Services.Tests.DriverServiceTests.CorrectDriverIsUsed("notest-assembly.dll",False,NUnit.Engine.Drivers.NUnit3FrameworkDriver)
  Expected: instance of <NUnit.Engine.Drivers.NUnit3FrameworkDriver>
  But was:  <NUnit.Engine.Drivers.InvalidAssemblyFrameworkDriver>
   at NUnit.Engine.Services.Tests.DriverServiceTests.CorrectDriverIsUsed(String fileName, Boolean skipNonTestAssemblies, Type expectedType) in C:\projects\nunit-console\src\NUnitEngine\nunit.engine.tests\Services\DriverServiceTests.cs:line 73
2) Failed : NUnit.Engine.Services.Tests.DriverServiceTests.CorrectDriverIsUsed("notest-assembly.dll",True,NUnit.Engine.Drivers.SkippedAssemblyFrameworkDriver)
  Expected: instance of <NUnit.Engine.Drivers.SkippedAssemblyFrameworkDriver>
  But was:  <NUnit.Engine.Drivers.InvalidAssemblyFrameworkDriver>
   at NUnit.Engine.Services.Tests.DriverServiceTests.CorrectDriverIsUsed(String fileName, Boolean skipNonTestAssemblies, Type expectedType) in C:\projects\nunit-console\src\NUnitEngine\nunit.engine.tests\Services\DriverServiceTests.cs:line 73
```

rprouse commented:
@ChrisMaddock you thought we should take this in stages. Based on that, this is ready for review for the first stage. Once this is complete, I will move on to merging .NET Standard solution into the main solution.

Can you review? Or @jnm2 if you have time.

ChrisMaddock commented:
Sure - travelling at the moment but should be able to take a look tomorrow. 😊

ChrisMaddock commented on src/NUnitEngine/nunit.engine.tests/nunit.engine.tests.csproj:
```diff
@@ -1,183 +1,54 @@
-    </FileUpgradeFlags>
-    <OldToolsVersion>3.5</OldToolsVersion>
-    <UpgradeBackupLocation />
-    <LangVersion>6</LangVersion>
```
Purposefully removed? Fine by me, but don't want it to be missed! Same applies for all csproj's

ChrisMaddock commented on src/NUnitEngine/nunit.engine.tests/nunit.engine.tests.csproj:
```diff
@@ -1,183 +1,54 @@
-  <PropertyGroup>
     <AssemblyOriginatorKeyFile>..\..\nunit.snk</AssemblyOriginatorKeyFile>
+    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+    <DebugType>Full</DebugType>
```
Would be nice but non essential to move all this signing and warningsaserrors stuff to a `Directory.Build.props` one day

ChrisMaddock replied:
This would also mean we could add the warnings-as-errors-but-not-in-VS that Joseph put in the framework project. But again - this is probably a separate PR for a separate day - I'm just getting over-excited! 😄

jnm2 replied:
😃 That was actually Sam! But yeah, very easy to do later, maybe better to do it later to to be more granular with our changes. Since this PR is about merging the projects, I'd expect mostly just csproj changes that making the merging easier.

jnm2 replied:
Do we still need `<DebugType>full</DebugType>`?

ChrisMaddock commented on src/NUnitEngine/nunit.engine.tests/nunit.engine.tests.csproj:
```diff
@@ -1,183 +1,54 @@
-      <HintPath>..\..\..\packages\NUnit.3.9.0\lib\net35\nunit.framework.dll</HintPath>
-    </Reference>
+  <ItemGroup Condition="'$(TargetFramework)'=='net35'">
+    <PackageReference Include="Mono.Cecil" Version="0.9.6.1" />
```
Why does this project reference Cecil directly? The only reason I spotted this was as it used to reference the net20 build specifically, and now references net35. I haven't looked if the both resolve to the same assembly - more wondering if this reference is actually necessary, and if it needs to be in sync with the version that the engine is using?

jnm2 commented on build.cake:
```diff
@@ -134,13 +134,9 @@ Task("InitializeBuild")
-            Source = PACKAGE_SOURCE,
-            Verbosity = NuGetVerbosity.Detailed
-        });
+		DotNetCoreRestore(SOLUTION_FILE);
```
I never really liked waiting for a custom Cake bootstrapper to download .NET locally. My personal preference would be for the Cake script to check `dotnet --version` and if the number is too low or `dotnet` does not exist, throw an exception with the minimum version required and the link to https://www.microsoft.com/net/download. This way it uses the shared SDK on your machine.
This is also how Roslyn does it, come to think of it.

jnm2 commented:
@rprouse 

> Can you review? Or @jnm2 if you have time.

I'm making you review a ton of large PRs, so it's only fair! 😁

jnm2 commented on src/NUnitEngine/mock-assembly/mock-assembly.csproj:
```diff
@@ -1,114 +1,30 @@
-    <Compile Include="AccessesCurrentTestContextDuringDiscovery.cs" />
+  <ItemGroup Condition="'$(TargetFramework)'=='net20'">
+    <Reference Include="System.Data" />
+    <Reference Include="System.Xml" />
```
I believe these assembly references are implicit with the new csproj for all .NET Framework targets.

jnm2 commented on src/NUnitEngine/mock-assembly/mock-assembly.csproj:
```diff
@@ -1,114 +1,30 @@
-    <RegisterForComInterop>false</RegisterForComInterop>
-    <RemoveIntegerChecks>false</RemoveIntegerChecks>
-    <WarningLevel>4</WarningLevel>
-    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
```
I think we still want `TreatWarningsAsErrors`, right? What about `RemoveIntegerChecks`?

jnm2 commented on src/NUnitEngine/nunit.engine.api/nunit.engine.api.csproj:
```diff
@@ -1,110 +1,23 @@
-    <ErrorReport>prompt</ErrorReport>
-    <WarningLevel>4</WarningLevel>
-    <DocumentationFile>$(OutputPath)nunit.engine.api.xml</DocumentationFile>
-    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
```
I think we still want `TreatWarningsAsErrors`, and a `<GenerateDocumentationFile>true</GenerateDocumentationFile>`, right?

jnm2 commented on src/NUnitEngine/nunit.engine.api/nunit.engine.api.csproj:
```diff
@@ -1,110 +1,23 @@
-  <PropertyGroup>
     <AssemblyOriginatorKeyFile>..\..\nunit.snk</AssemblyOriginatorKeyFile>
+    <DocumentationFile>$(OutputPath)\$(TargetFramework)\nunit.engine.api.xml</DocumentationFile>
+    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
```
Ah, found it!  `<GenerateDocumentationFile>true</GenerateDocumentationFile>` is what the framework projects use now, btw.

jnm2 commented on src/NUnitEngine/nunit.engine.api/nunit.engine.api.csproj:
```diff
@@ -1,110 +1,23 @@
     <Reference Include="System" />
     <Reference Include="System.Data" />
     <Reference Include="System.Drawing" />
     <Reference Include="System.Xml" />
```
I think all four of these are implicit for .NET Framework, so the whole ItemGroup could disappear.

jnm2 commented on src/NUnitEngine/nunit.engine/nunit.engine.csproj:
```diff
@@ -1,183 +1,35 @@
     <Reference Include="System.Drawing" />
     <Reference Include="System.Runtime.Remoting" />
     <Reference Include="System.Web" />
     <Reference Include="System.Xml" />
```
I think `System`, `System.Data`, `System.Drawing` and `System.Xml` are implicit.

jnm2 commented on src/NUnitEngine/nunit.engine/nunit.engine.csproj:
```diff
@@ -12,12 +12,12 @@
     <Reference Include="System.Drawing" />
     <Reference Include="System.Runtime.Remoting" />
     <Reference Include="System.Web" />
     <Reference Include="System.Xml" />
```
Some implicit.

jnm2 commented on src/NUnitConsole/nunit3-console.tests/nunit3-console.tests.csproj:
```diff
@@ -1,78 +1,26 @@
+    <OutputPath>..\..\..\bin\$(Configuration)\</OutputPath>
     <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
     <NoWarn>1685</NoWarn>
+    <DebugType>Full</DebugType>
```
Still need? (here and other places)

jnm2 replied:
I didn't think the engine's usage of Cecil caused the PDB to be read the way the adapter did. At least, I'm using NUnit Console in all my build scripts and don't have <DebugType> set anymore. Maybe NUnit Console is gracefully losing some info like line numbers?

jnm2 commented on src/NUnitConsole/nunit3-console.tests/nunit3-console.tests.csproj:
```diff
@@ -1,78 +1,26 @@
+  <ItemGroup Condition="'$(TargetFramework)'=='net35'">
     <Reference Include="System" />
     <Reference Include="System.Data" />
     <Reference Include="System.Xml" />
```
Implicit.

jnm2 commented on src/NUnitConsole/nunit3-console/nunit3-console.csproj:
```diff
@@ -1,116 +1,33 @@
+  <ItemGroup Condition="'$(TargetFramework)'=='net20'">
     <Reference Include="System" />
     <Reference Include="System.Data" />
     <Reference Include="System.Xml" />
```
Implicit.

jnm2 commented on src/NUnitConsole/nunit3-console/nunit3-console.csproj:
```diff
@@ -1,116 +1,33 @@
+    <Compile Include="..\..\CommonAssemblyInfo.cs" Link="Properties\CommonAssemblyInfo.cs" />
+    <Compile Include="..\ConsoleVersion.cs" Link="Properties\ConsoleVersion.cs" />
+    <Compile Include="..\..\NUnitEngine\nunit.engine\Guard.cs" Link="Utilities\Guard.cs" />
+    <Compile Include="..\..\NUnitEngine\nunit.engine\Internal\ExceptionHelper.cs" Link="Utilities\ExceptionHelper.cs" />
```
Side note, I learned that these can be written as `LinkBase="Properties"` etc.

jnm2 commented on src/NUnitConsole/nunit3-console/nunit3-console.csproj:
```diff
@@ -1,116 +1,33 @@
+    <Compile Include="..\..\NUnitEngine\nunit.engine\Internal\ExceptionHelper.cs" Link="Utilities\ExceptionHelper.cs" />
+    <Content Include="..\..\..\nunit.ico" Link="nunit.ico" />
     <None Include="app.config" />
     <None Include="app.manifest" />
```
I think all files show up as None implicitly.

jnm2 commented on src/NUnitEngine/nunit-agent/nunit-agent-x86.csproj:
```diff
@@ -1,90 +1,35 @@
     <Reference Include="System" />
     <Reference Include="System.Runtime.Remoting" />
-    <Reference Include="System.Windows.Forms" />
     <Reference Include="System.Xml" />
```
Implicit.

jnm2 commented on src/NUnitEngine/nunit-agent/nunit-agent-x86.csproj:
```diff
@@ -1,90 +1,35 @@
   </ItemGroup>
   <ItemGroup>
     <None Include="app.config" />
     <None Include="app.manifest" />
```
Implicit.

jnm2 commented on src/NUnitEngine/nunit-agent/nunit-agent.csproj:
```diff
@@ -1,87 +1,33 @@
+  <ItemGroup Condition="'$(TargetFramework)'=='net20'">
     <Reference Include="System" />
     <Reference Include="System.Runtime.Remoting" />
     <Reference Include="System.Xml" />
```
Implicit.

jnm2 commented on src/NUnitEngine/nunit-agent/nunit-agent.csproj:
```diff
@@ -1,87 +1,33 @@
   </ItemGroup>
   <ItemGroup>
     <None Include="app.config" />
     <None Include="app.manifest" />
```
Implicit.

jnm2 commented on build.cake:
```diff
@@ -25,7 +25,7 @@ var packageVersion = version + modifier + dbgSuffix;
 var PROJECT_DIR = Context.Environment.WorkingDirectory.FullPath + "/";
 var PACKAGE_DIR = PROJECT_DIR + "package/";
-var BIN_DIR = PROJECT_DIR + "bin/" + configuration + "/";
+var BIN_DIR = PROJECT_DIR + "bin/" + configuration + "/net35/";
```
@rprouse Does CreateImage include the .NET Standard version of the engine?

ChrisMaddock replied:
I think the test projects were upgraded to .NET 3.5 to support NSubstitute.

ChrisMaddock replied:
It makes me a little nervous copying a dll from the `net35` directory and packaging it under `net20` - but it would take a fair bit of work to redo, and probably isn't worth the effort. 🙂

jnm2 commented on .travis.yml:
```diff
@@ -22,4 +22,4 @@ matrix:
 
 script:
-  - ./build.sh --target=Travis --configuration=Release
+  - ./build.sh --target "Travis" --configuration "Release"
```
😢 This comes down to which `.sh` bootstrapper you're using.

jnm2 commented on build.cake:
```diff
@@ -134,13 +134,9 @@ Task("InitializeBuild")
+		DotNetCoreRestore(SOLUTION_FILE);
 
-        if(IsDotNetCoreInstalled && IsRunningOnWindows())
+        if(IsRunningOnWindows())
```
Why only on Windows? Does it have to do with `DotNetCoreRestore`? I only use MSBuild 15 and `.WithTarget("Restore")` these days, and that's what the framework is doing.

rprouse commented:
@jnm2 are there additional changes you would like on this, or shall we merge and get on to stage 2?
